### PR TITLE
Discuss rules

### PR DESCRIPTION
**Pull request by joao-paulo-parity**
[Rendered](https://github.com/paritytech/pr-custom-review/blob/rules/rules.md)

---

Given the reasoning in https://github.com/paritytech/substrate/pull/10951#issuecomment-1072283536, I plan to use this pull request exclusively for discussing repository rules. It will not be merged.

Once a consensus has been reached, we'll implement what is defined in this PR without having to discuss again for each repository.

The plan for this PR is as follows:

Step 1: I'll collect the requirements with @TriplEight 
Step 2: We'll invite one or more stakeholders to clarify *specific* details related to their expertise
Step 3: We'll request reviews from **all** stakeholders

#32, #35, #36, #38 can be closed after the discussion is finished

### CODE REVIEW

File: rules.md
Comment (joao-paulo-parity) [inline rules.md:12]: Guess based on https://github.com/paritytech/substrate/pull/10951#pullrequestreview-908761989 but I need confirmation

File: rules.md
Comment (joao-paulo-parity) [inline rules.md:19]: Guess based on https://github.com/paritytech/substrate/pull/10951#pullrequestreview-908761989 but I need confirmation

File: rules.md
Comment (joao-paulo-parity) [inline rules.md:19]: Guess based on https://github.com/paritytech/substrate/pull/10951#pullrequestreview-908761989 but needs confirmation

File: rules.md
Comment (joao-paulo-parity) [inline rules.md:19]: Guess based on https://github.com/paritytech/substrate/pull/10951#pullrequestreview-908761989 but needs confirmation

Comment (TriplEight) [inline rules.md:19]:
```suggestion
  - `polkadot-parachains/{statemine|statemint}/src/*.rs`, except weight files
```

File: rules.md
Comment (TriplEight) [reply rules.md:12]:
```suggestion
  - `/runtime/{kusama|polkadot}/src/*.rs`, except weight files
```

**Review by TriplEight — CHANGES_REQUESTED** [review]

**Review by TriplEight — DISMISSED** [review]

File: rules.md
Comment (bkchr) [inline rules.md:3]:
```suggestion
- 🔒️ed lines are reviewed by 2 team leads, one of them must be in the group {@gavofyork, @shawntabrizi, @rphmeier, @bkchr}
```

Comment (bkchr) [reply rules.md:3]: These changes including pinging the people?

File: rules.md
Comment (bkchr) [inline rules.md:4]: This does not include pinging.

File: rules.md
Comment (bkchr) [inline rules.md:8]:
```suggestion
- 🔒️ed lines are reviewed by 2 team leads, one of them must be in the group {@gavofyork, @shawntabrizi, @rphmeier, @bkchr}
```

Comment (bkchr) [reply rules.md:8]: Includes pinging.

File: rules.md
Comment (bkchr) [inline rules.md:9]: Includes no pinging.

File: rules.md
Comment (bkchr) [inline rules.md:10]:
```suggestion
- The following files are reviewed by 2 team leads, one of them must be in the group {@gavofyork, @shawntabrizi, @rphmeier, @bkchr}
```

Comment (bkchr) [reply rules.md:10]: Includes pinging.

File: rules.md
Comment (bkchr) [inline rules.md:15]:
```suggestion
- 🔒️ed lines are reviewed by 2 team leads, one of them must be in the group {@gavofyork, @joepetrowski, @shawntabrizi, @rphmeier, @bkchr}
```

Comment (bkchr) [reply rules.md:15]: Includes pinging.

File: rules.md
Comment (bkchr) [inline rules.md:16]: Includes no pinging.

File: rules.md
Comment (bkchr) [inline rules.md:17]:
```suggestion
- The following files are reviewed by 2 team leads, one of them must be in the group {@gavofyork, @joepetrowski, @shawntabrizi, @rphmeier, @bkchr}
```

Comment (bkchr) [reply rules.md:17]: Includes pinging.

**Review by bkchr — COMMENTED** [review]
For the team leads group, please create an empty team on github and then I will work with @dvdplm to fill this team :)

**Comment by TriplEight** [comment]
@bkchr 

> For the team leads group, please create an empty team on github and then I will work with @dvdplm to fill this team :)

Here you go: https://github.com/orgs/paritytech/teams/parachainleads

File: rules.md
Comment (TriplEight) [reply rules.md:3]: yes, as we've discussed
```suggestion
- 🔒️ed lines are reviewed by 2 team leads, one of them must be in the group {@gavofyork, @shawntabrizi, @rphmeier, @bkchr} and the other in another team, @paritytech/polkadotleads 
```

File: rules.md
Comment (TriplEight) [reply rules.md:4]: as per our discussion, this can be done by **not assigning** the team, but _passively-aggressively waiting and failing a job until someone suitable reviews a pr_. Details may be shown in the job's output.

File: rules.md
Comment (TriplEight) [reply rules.md:8]: yes, as we've discussed
```suggestion
- 🔒️ed lines are reviewed by 2 team leads, one of them must be in the group {@gavofyork, @shawntabrizi, @rphmeier, @bkchr} and the other in another team, @paritytech/polkadotleads 
```

File: rules.md
Comment (TriplEight) [reply rules.md:10]: yes, as we've discussed
```suggestion
- The following files are reviewed by 2 team leads, one of them must be in the group {@gavofyork, @shawntabrizi, @rphmeier, @bkchr} and the other in another team, @paritytech/polkadotleads 
```

File: rules.md
Comment (TriplEight) [reply rules.md:15]: yes, as we've discussed
```suggestion
- 🔒️ed lines are reviewed by 2 team leads, one of them must be in the group {@joepetrowski, @shawntabrizi, @rphmeier, @bkchr} and the other in another team, @paritytech/polkadotleads 
```

File: rules.md
Comment (TriplEight) [reply rules.md:17]: yes, as we've discussed
```suggestion
- The following files are reviewed by 2 team leads, one of them must be in the group {@joepetrowski, @shawntabrizi, @rphmeier, @bkchr} and the other in another team, @paritytech/polkadotleads 
```

File: rules.md
Comment (TriplEight) [inline rules.md:3]:
```suggestion
- 🔒️ed lines are reviewed by 2 team leads, one of them must be in @paritytech/locks-review and the other in @paritytech/polkadotleads 
```

File: rules.md
Comment (TriplEight) [inline rules.md:8]:
```suggestion
- 🔒️ed lines are reviewed by 2 team leads, one of them must be in @paritytech/locks-review and the other in @paritytech/polkadotleads 
```

File: rules.md
Comment (TriplEight) [inline rules.md:10]:
```suggestion
- The following files are reviewed by 2 team leads, one of them must be in @paritytech/locks-review and the other in @paritytech/polkadotleads
```

File: rules.md
Comment (TriplEight) [inline rules.md:15]:
```suggestion
- 🔒️ed lines are reviewed by 2 team leads, one of them must be in @paritytech/cumulus-review and the other in @paritytech/polkadotleads
```

File: rules.md
Comment (TriplEight) [inline rules.md:17]:
```suggestion
- The following files are reviewed by 2 team leads, one of them must be in @paritytech/cumulus-review and the other in @paritytech/polkadotleads
```

**Comment by TriplEight** [comment]
to summarize here what we've discussed with Basti:

There are now 3 new groups:
`@paritytech/locks-review` `@paritytech/cumulus-review` and `@paritytech/parachainleads` (yet to be populated).

- the rules about :lock: s should ping 2 entire groups: `@paritytech/parachainleads` AND one of `@paritytech/cumulus-review` OR `@paritytech/locks-review` depending on the repo
- the rules about changed runtime files should ping 2 entire groups: `@paritytech/parachainleads` AND one of `@paritytech/cumulus-review` OR `@paritytech/locks-review` depending on the repo
- the rest regular PRs are reviewed by `@paritytech/core-devs` without pinging them. I assume this can be done by **not assigning the team**, but _passively-aggressively waiting and failing a job until someone suitable reviews a pr_. Details may be shown in the job's output.

File: rules.md
Comment (bkchr) [inline rules.md:3]: `and the other in @paritytech/polkadotleads `

It could still be that the second reviewer is also in locks-review. As locks-review is a subset of the other group.

File: rules.md
Comment (bkchr) [inline rules.md:15]: Maybe call it `cumulus-locks-review`

**Comment by bkchr** [comment]
Otherwise the rules are looking good and are also approved by @gavofyork

**Comment by gavofyork** [comment]
Looks reasonable to me.

File: rules.md
Comment (bkchr) [reply rules.md:3]: > Basically, the corner case here would be not to accept reviews as sufficient when there are two reviews from `@paritytech/polkadotleads` and none from `@paritytech/locks-review`

I have seen the other case as the corner case, but in the end it isn't that important as long as both cases are ensured :)

File: rules.md
Comment (TriplEight) [inline rules.md:10]:
```suggestion
- The following files are reviewed by 2 team leads, one of them must be in @paritytech/locks-review and the other in @paritytech/parachainleads
```

File: rules.md
Comment (TriplEight) [inline rules.md:15]:
```suggestion
- 🔒️ed lines are reviewed by 2 team leads, one of them must be in @paritytech/cumulus-review and the other in @paritytech/parachainleads
```

File: rules.md
Comment (TriplEight) [inline rules.md:17]:
```suggestion
- The following files are reviewed by 2 team leads, one of them must be in @paritytech/cumulus-review and the other in @paritytech/parachainleads
```

**Review by TriplEight — CHANGES_REQUESTED** [review]

**Comment by TriplEight** [comment]
Here are the new teams' names we've agreed upon w Basti
image

File: rules.md
Comment (joao-paulo-parity) [inline rules.md:4]: Replace "be fail" with "fail" everywhere

Replace "This group should not be explicitly mentioned" with "This team's reviews should not be requested"

So I take we'll not be doing https://docs.github.com/en/organizations/organizing-members-into-teams/managing-code-review-settings-for-your-team#configuring-auto-assignment.

Comment (joao-paulo-parity) [reply rules.md:4]: I might have misinterpreted "mentioned" (pr-custom-review doesn't mention anything) since I assumed it's "requested". Is this related to https://github.com/paritytech/substrate/pull/10951#discussion_r827314600? There I explained that not everyone should be pinged for each PR given that https://docs.github.com/en/organizations/organizing-members-into-teams/managing-code-review-settings-for-your-team#configuring-auto-assignment is properly configured.

Comment (TriplEight) [reply rules.md]: > So I take we'll not be doing https://docs.github.com/en/organizations/organizing-members-into-teams/managing-code-review-settings-for-your-team#configuring-auto-assignment for `core-devs`

no, at least not yet. `core-devs` is a very large group (and for me, it's still a question why wouldn't we create more narrow-specialized groups. But it's surely OK to start like this). And it doesn't really make sense to either ping them all or do the auto-assignment in that group.

File: rules.md
Comment (dvdplm) [inline rules.md:4]:
```suggestion
- PRs that do not touch locked lines require 2 positive reviews from @paritytech/core-devs.
```

I think specifying the notification strategy is an implementation detail and not part of the rules. Same for the messaging on failure: a requirement for the implementation but not really part of the rule. 

Any failure should explain itself and with the current wording it almost sounds like the locked lines rule does not need to explain anything, while the normal-PR-rule does require it.

File: rules.md
Comment (dvdplm) [inline rules.md:3]: I know this was covered in the discussion, but perhaps it's worth clarifying what this rule means for leads present in both `locks-review` and `polkadot-review` teams?

File: rules.md
Comment (joao-paulo-parity) [reply rules.md]: > and for me, it's still a question why wouldn't we create more narrow-specialized groups

I can see that being reasonable given that core-devs will not be a blocker for processbot anymore after pr-custom-review is applied.

> it doesn't really make sense to either ping them all or do the auto-assignment in that group

Alright

**Comment by joao-paulo-parity** [comment]
FYI I plan to come back to this ASAP which is probably tomorrow or early next week. I've identified https://github.com/paritytech/pr-custom-review/issues/68, https://github.com/paritytech/pr-custom-review/issues/69, https://github.com/paritytech/pr-custom-review/issues/70 and https://github.com/paritytech/pr-custom-review/issues/71 as tasks to be worked on for us to support the requirements discussed here.

**Comment by joao-paulo-parity** [comment]
The preliminary PRs for the configuration have been submitted.

The application of those configurations through the GitHub workflow should be coordinated with a simultaneous update of processbot (https://github.com/paritytech/parity-processbot/pull/373) since the current requirements are conflicting with processbot's own review requirements.

File: rules.md
Comment (joao-paulo-parity) [inline rules.md:3]: It seems like there's some disagreement about this requirement and others similar to it. I'll try to decompose the sentence according to my literal interpretation:

> 🔒️ed lines are reviewed by **2** team leads

Emphasis on "2".

> one of them must be in ...

Let's call this agent `DeveloperA`.

> and the other in ...

"the other" literally refers to another agent different from `DeveloperA`. Let's call this `DeveloperB`.

All of the above phrasing to me strongly suggests 2 different agents for this requirement.

Now it might be that someone is in both `@paritytech/locks-review` and `@paritytech/polkadot-review`. https://matrix.to/#/!gJeGMHCcDoIwsHIJri:matrix.parity.io/$5H7WC-VGc4Am5HuvivqpuqA-jR_ic2SmWiz0x4HK1cg raised the question:

> why are you only counting towards one group if you are part of multiple? 

If we'd make it so 1 person can fulfill the criteria by being in both groups, we'd effectively lower the requirement to a single approval (1 person) instead of 2 like I understood by the literal interpretation. If that's intentional, I suggest clarifying the phrasing as e.g.

>🔒️ed requires 1 approval from @paritytech/locks-review and 1 approval from @paritytech/polkadot-review

That removes any implication of "2 team leads" since it only mentions 1 approval from each group, which might be given by the same person if they are in both groups. What's your opinion?

cc @bkchr @TriplEight

Comment (joao-paulo-parity) [inline rules.md:3]: It seems like there's some disagreement about this requirement and others similar to it. I'll try to decompose the sentence according to my literal interpretation:

> 🔒️ed lines are reviewed by **2** team leads

Emphasis on "2".

> one of them must be in ...

Let's call this agent `DeveloperA`.

> and the other in ...

"the other" literally refers to another agent different from `DeveloperA`. Let's call this `DeveloperB`.

All of the above phrasing to me strongly suggests 2 different agents for this requirement.

Now it might be that someone is in both `@paritytech/locks-review` and `@paritytech/polkadot-review`. https://matrix.to/#/!gJeGMHCcDoIwsHIJri:matrix.parity.io/$5H7WC-VGc4Am5HuvivqpuqA-jR_ic2SmWiz0x4HK1cg raised the question:

> why are you only counting towards one group if you are part of multiple? 

If we'd make it so 1 person can fulfill the criteria by being in both groups, we'd effectively lower the requirement to a single approval (1 person) instead of 2 like I understood by the literal interpretation. If that's intentional, I suggest clarifying the phrasing as e.g.

>🔒️ed requires 1 approval from @paritytech/locks-review and 1 approval from @paritytech/polkadot-review

That removes any implication of "2 team leads" since it only mentions 1 approval from each group, which might be given by the same person if they are in both groups. What's your opinion?

cc @bkchr @TriplEight

Comment (TriplEight) [reply rules.md:3]: I agree that 1 person that is a member of two teams shall count as a review from both teams. But also it should count as **one** review of two requested.

Comment (joao-paulo-parity) [reply rules.md:3]: > A second team lead from either team is then required to make the check succeed.

"Either" team? That's different from how I was understanding your suggestion.

I think it'll help to walk through an example scenario. Suppose the following structure

```
├── locks-review
│  ├── user1
│  └── user2
├── polkadot-review
│  ├── user1
│  └── user3
```

**Alternative 1** (how it works currently): after `user1` approves we'll still need an approval from `user2` or `user3`, even though `user1` is in both teams.

Suggested wording:

> :lock: lines requires 1 approval from @paritytech/locks-review and 1 approval from @paritytech/polkadot-review. Each approval should come from a different user.

Note: I think adding "Each approval should come from a different user" is clearer than saying "2 team leads".

---

**Alternative 2** (what I _**initially**_ understood from your suggestion): `user1`'s approval would fulfill the requirement because they are on both teams, therefore not requiring an approval from `user2` or `user3`.

Suggested wording:

> :lock: lines requires 1 approval from @paritytech/locks-review and 1 approval from @paritytech/polkadot-review

Note: avoids mentioning "2 team leads" in order to not provoke any ambiguity, since a single team lead is enough to fulfill both requirements. Also avoids mentioning "Each approval should come from a different user" since it's intentionally not that way.

---

**Alternative 3** (how I understood from "either"): after `user1` approves we'll need an approval from **either** `user2` or `user3`.

Suggested wording:

> :lock: lines requires 2 approvals from users on either @paritytech/locks-review or @paritytech/polkadot-review. Each approval should come from a different user.

---

Which alternative are you suggesting? Also feel free to describe another alternative if none is matching

Comment (bkchr) [reply rules.md:3]: Isn't alternative 1 and 3 the same? 

And is alternative 1 really how it currently works? I doubt this.

Comment (bkchr) [reply rules.md:3]: Okay, I think I may mixed something up. I'm now confused what the current status is. I just know that yesterday it wasn't alternative 1.

Comment (bkchr) [reply rules.md:3]: I have now extended the polkadot-review team and now it should be okay.

Comment (joao-paulo-parity) [reply rules.md:3]: > Isn't alternative 1 and 3 the same?

In this specific example, when `user1` approves, they *result* in the same outcome but their definitions are still  different.

Consider if `user2` were to approve. "Alternative 1" in that case would require that other approval comes from specifically `polkadot-review` (`user2` isn't a part of it). However, if you say "total approvals from either team" then the second approval could come from the same team that `user2` is in. In this reduced example with a few team members they would might effectively result in the same conditions, but I think it would be different if you start adding more team members like the groups we have currently.

> I just know that yesterday it wasn't alternative 1

If it's about https://github.com/paritytech/polkadot/runs/5978667098?check_suite_focus=true#step:2:43 it was misbehaving due to a bug, but it should be fine now.

Comment (bkchr) [reply rules.md:3]: Then let's take alternative 1

Comment (jsdw) [reply rules.md:3]: FWIW I thought that the initial "🔒️ed lines are reviewed by 2 team leads, one of them must be in @paritytech/locks-review and the other in @paritytech/polkadot-review." was precise enough to imply alternative 1 as-is, but if that is unclear, I'd go for something like:

> 🔒️ed lines must be reviewed by 2 team leads. One of them must be a member of @paritytech/locks-review and the other must be a member of @paritytech/polkadot-review.

If that is _still_ unclear, then I'd add somehting like @bkchr said:

> 🔒️ed lines must be reviewed by 2 team leads. One of them must be a member of @paritytech/locks-review and the other must be a member of @paritytech/polkadot-review. If one is a member of both teams, then the other may be a member of either team.

**Pull request closed without merging**
